File: python/physics/bounce/physics.py
"""Ball physics — no rendering dependency.

Contains the two functions worth benchmarking: ``update`` (position +
wall reflection) and ``speed`` (scalar magnitude). Both a pure-Python
and a numpy-vectorised variant are provided so the benchmark can show
the difference.
"""

import math
from dataclasses import dataclass


@dataclass
class Ball:
    """Mutable state for a single bouncing ball.

    Attributes:
        x:  Centre x position (pixels).
        y:  Centre y position (pixels).
        vx: Horizontal velocity (pixels/frame).
        vy: Vertical velocity (pixels/frame).
    """

    x: float
    y: float
    vx: float
    vy: float


def update(ball: Ball, radius: int, width: int, height: int) -> None:
    """Advance ball by one frame; reflect elastically off walls.

    Modifies ball in place. Position is updated first, then each velocity
    component is negated when the ball edge crosses a boundary.

    Args:
        ball:   Ball state, modified in place.
        radius: Ball radius in pixels (must be > 0).
        width:  Window width in pixels.
        height: Window height in pixels.
    """
    ball.x += ball.vx
    ball.y += ball.vy
    if ball.x - radius < 0 or ball.x + radius > width:
        ball.vx = -ball.vx
    if ball.y - radius < 0 or ball.y + radius > height:
        ball.vy = -ball.vy


def speed(ball: Ball) -> float:
    """Return the scalar speed of the ball in pixels/frame.

    Args:
        ball: Ball state (read-only).

    Returns:
        sqrt(vx² + vy²).
    """
    return math.sqrt(ball.vx ** 2 + ball.vy ** 2)

File: nodejs/physics/bounce/src/physics.ts
/**
 * @file   physics.ts
 * @brief  Ball physics — no rendering dependency.
 *
 * Contains the two functions worth benchmarking: `update` (position +
 * wall reflection) and `speed` (scalar magnitude).
 */

/** Mutable state for a single bouncing ball. */
export interface Ball {
    x:  number;   // Centre x position (pixels)
    y:  number;   // Centre y position (pixels)
    vx: number;   // Horizontal velocity (pixels/frame)
    vy: number;   // Vertical velocity (pixels/frame)
}

/**
 * Advances the ball by one frame and reflects elastically off walls.
 *
 * Mutates the ball object in place. Position is updated first, then each
 * velocity component is negated when the ball edge crosses a boundary.
 *
 * @param ball   Ball state, mutated in place.
 * @param radius Ball radius in pixels (must be > 0).
 * @param width  Window width in pixels.
 * @param height Window height in pixels.
 */
export function update(ball: Ball, radius: number, width: number, height: number): void {
    ball.x += ball.vx;
    ball.y += ball.vy;
    if (ball.x - radius < 0 || ball.x + radius > width)  ball.vx = -ball.vx;
    if (ball.y - radius < 0 || ball.y + radius > height) ball.vy = -ball.vy;
}

/**
 * Returns the scalar speed of the ball in pixels/frame.
 *
 * @param ball Ball state (read-only).
 * @returns sqrt(vx² + vy²).
 */
export function speed(ball: Ball): number {
    return Math.sqrt(ball.vx ** 2 + ball.vy ** 2);
}
